services/bonita-business-data/bonita-business-data-impl/src/main/java/org/bonitasoft/engine/business/data/impl/BusinessDataServiceImpl.java

```diff
@@ -333,6 +334,27 @@ public BusinessDataQueryResult getJsonQueryEntities(final String entityClassName
         return new BusinessDataQueryResultImpl(buildJsonRepresentation((List<Entity>) list, businessDataURIPattern), businessDataQueryMetadata);
     }
 
+    private Query getCountQueryDefinition(Class<? extends Entity> businessDataClass, BusinessObject businessObject, Query queryDefinition) {
+        final Query countQueryDefinition = countQueryProvider.getCountQueryDefinition(businessObject, queryDefinition);
+        if (ensureQueryIsDefinedInEntity(businessDataClass, countQueryDefinition)) {
+            return countQueryDefinition;
+        }
+        return null;
+    }
+
+    private boolean ensureQueryIsDefinedInEntity(Class<? extends Entity> businessDataClass, Query countQueryDefinition) {
+        final NamedQueries namedQueries = businessDataClass.getAnnotation(NamedQueries.class);
+        if (namedQueries == null || countQueryDefinition == null) {
+            return false;
+        }
+        for (NamedQuery namedQuery : namedQueries.value()) {
+            if (namedQuery.name().equals(getQualifiedQueryName(businessDataClass, countQueryDefinition.getName()))) {
```

Comment:
Yes in fact it is here that we can use BDMQueryUtil.getCountQueryName(baseQuery.getName())
instead of calling all the things in CountQueryProvider that gather every generated queries and so on, even if it works like that I think we should change that.
appart that LGTM :)